src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java

```diff
@@ -168,56 +194,147 @@ public AssemblyResultSet runLocalAssembly(final AssemblyRegion assemblyRegion,
+                    assemblyResultByRTGraph.put(assembledResult.getThreadingGraph(), assembledResult);
+                    nonRefRTGraphs.add(assembledResult.getThreadingGraph());
+
+                    if (graphHaplotypeHistogramPath != null) {
+                        kmersUsedHistogram.add((double) assembledResult.getKmerSize());
+                    }
+
+                    AbstractReadThreadingGraph graph = assembledResult.getThreadingGraph();
+                    findBestPaths(Collections.singletonList(graph), Collections.singletonMap(graph, assembledResult),
+                            refHaplotype, refLoc, activeRegionExtendedLocation, null, aligner);
+
+                    savedAssemblyResults.add(assembledResult);
+
+                    //TODO LOGIC PLAN HERE - we want to check if we have a trustworthy graph (i.e. no badly assembled haplotypes) if we do, emit it.
+                    //TODO                 - but if we failed to assemble due to excessive looping or did have badly assembled haplotypes then we expand kmer size.
+                    //TODO                 - If we get no variation
+
+                    // if asssembly didn't fail ( which is a degenerate case that occurs for some subset of graphs with difficult loops)
+                    if (! savedAssemblyResults.get(savedAssemblyResults.size() - 1).getHaplotypeList().isEmpty()) {
+                        // we have found our workable kmer size so lets add the results and finish
+                        if (!assembledResult.isContainsSuspectHaplotypes()) {
+                            for (Haplotype h : assembledResult.getHaplotypeList()) {
+                                resultSet.add(h, assembledResult);
+                            }
+                            hasAdequatelyAssembledGraph = true;
+                        }
+                    }
+
+                // if no variation is discoverd in the graph don't bother expanding the kmer size.
+                } else if (assembledResult != null && assembledResult.getStatus() == AssemblyResult.Status.JUST_ASSEMBLED_REFERENCE) {
+                    hasAdequatelyAssembledGraph = true;
+                }
+            }
         }
 
-        // print the graphs if the appropriate debug option has been turned on
-        if ( graphOutputPath != null ) { printGraphs(nonRefSeqGraphs); }
-        if ( graphHaplotypeHistogramPath != null ) { haplotypeHistogram.add((double)resultSet.getHaplotypeCount()); }
 
-        return resultSet;
+        // This indicates that we have thrown everything away... we should go back and check that we weren't too conservative about assembly results that might otherwise be good
+        if (!hasAdequatelyAssembledGraph) {
+            // search for the last haplotype set that had any results, if none are found just return me
+            // In this case we prefer the last meaningful kmer size if possible
+            for (int i = savedAssemblyResults.size() - 1; i > 0; i --) {
```

Comment:
How about `for savedAssemblyResult : Lists.reverse(savedAssemblyResults)`, which iterates over a reversing wrapper of the list?